File: backend/schemas/device.py
# backend/schemas/device.py
from pydantic import BaseModel, ConfigDict, IPvAnyAddress, computed_field
from datetime import datetime
from typing import Optional

class DeviceResponse(BaseModel):
    id: int
    hostname: str
    ip_address: str  # cambiado de IPvAnyAddress a str para evitar problemas de serialización
    site: Optional[str] = None
    version: Optional[str] = None

    # --- NUEVOS FACTS ---
    platform: Optional[str] = None
    serial_number: Optional[str] = None
    ha_state: Optional[str] = None          # ACTIVE | STANDBY | ...
    sync_status: Optional[str] = None       # In Sync | Changes Pending | ...
    last_sync_color: Optional[str] = None   # green | yellow | red
    dns_servers: Optional[str] = None       # CSV o JSON string (como lo guardas)
    last_facts_refresh: Optional[datetime] = None

    # --- SCAN INFO ---
    last_scan_status: Optional[str] = None
    last_scan_message: Optional[str] = None 
    last_scan_timestamp: Optional[datetime] = None

    # --- Flags ---
    active: bool = True
    cluster_key: Optional[str] = None
    is_primary_preferred: bool = False

    @computed_field
    @property
    def cluster_label(self) -> Optional[str]:
        if not self.cluster_key:
            return None
        parts = self.cluster_key.split("::", 1)
        return parts[1] if len(parts) == 2 else self.cluster_key

    # Campo computado para compatibilidad con la GUI (si espera last_sync)
    @computed_field
    @property
    def last_sync(self) -> Optional[str]:
        if self.last_scan_timestamp is None:
            return None
        return self.last_scan_timestamp.isoformat()

    # Campo computado para verificar si el dispositivo tiene credenciales
    has_credentials: bool = False

    model_config = ConfigDict(
        from_attributes=True,
        populate_by_name=True,
        extra='ignore'
    )

File: frontend/src/types/device.ts
// frontend/src/types/device.ts
// Tipos centralizados para Device - SINGLE SOURCE OF TRUTH

/**
 * Device completo con todos los campos del backend
 */
export interface Device {
  id: number;
  hostname: string;
  ip_address: string;
  site?: string | null;
  cluster_key?: string | null;
  is_primary_preferred: boolean;
  version?: string | null;
  platform?: string | null;
  serial_number?: string | null;
  ha_state?: string | null;          // ACTIVE | STANDBY | STANDALONE
  sync_status?: string | null;       // In Sync | Changes Pending | Disconnected
  last_sync_color?: string | null;   // green | yellow | red
  dns_servers?: string | null;
  last_facts_refresh?: string | null;
  last_scan_status?: string | null;  // success | error | running | pending
  last_scan_message?: string | null;
  last_scan_timestamp?: string | null;
  active: boolean;
  username?: string | null;
  created_at?: string;
  updated_at?: string;
}

/**
 * Alias para compatibilidad con código existente (DeviceRow = Device)
 */
export type DeviceRow = Device;

/**
 * Props for DeviceTable component
 */
export interface DeviceTableProps {
  onSetCredentials: (device: Device) => void;
  onDeleteDevice: (deviceId: number) => void;
  onRowClick?: (device: Device) => void;
  searchTerm?: string;
  refreshTrigger?: number;
  userRole?: string;
  onSelectionChange?: (ids: number[]) => void;
  clearSelectionKey?: number;
  filters?: DeviceFilters;
  onDevicesLoaded?: (devices: Device[]) => void;
  favorites?: number[];
  visibleColumns?: string[];
  onToggleFavorite?: (deviceId: number) => void;
}

/**
 * Device filters for DeviceTable
 */
export interface DeviceFilters {
  ha_state?: string | null;
  sync_status?: string | null;
  site?: string | null;
  is_primary_preferred?: boolean | null;
  no_credentials?: boolean | null;
  health_status?: string | null;
}

/**
 * Device mínimo para selects y listas simples
 */
export interface DeviceMinimal {
  id: number;
  hostname: string;
  ip_address?: string;
  site?: string | null;
}

/**
 * Credenciales de un device
 */
export interface DeviceCredentials {
  username: string;
  password: string;
}

/**
 * Payload para crear un device
 */
export interface DeviceCreate {
  hostname: string;
  ip_address: string;
  site?: string;
  cluster_key?: string;
  is_primary_preferred?: boolean;
  active?: boolean;
  username?: string;
  password?: string;
}

/**
 * Payload para actualizar un device
 */
export interface DeviceUpdate {
  hostname?: string;
  ip_address?: string;
  site?: string;
  cluster_key?: string;
  is_primary_preferred?: boolean;
  active?: boolean;
}

/**
 * Respuesta de scan/refresh
 */
export interface ScanResponse {
  queued: boolean;
  message?: string;
}
